### PR TITLE
Adding a parents helper API to MerkleReg

### PR DESCRIPTION
**Pull request by bochaco**

### CODE REVIEW

File: src/merkle_reg.rs [hunk 1]
```diff
@@ -1,8 +1,8 @@
 use core::convert::Infallible;
 use core::fmt;
-use std::collections::{BTreeMap, BTreeSet};
-
 use quickcheck::{Arbitrary, Gen};
+use serde::{Deserialize, Serialize};
+use std::collections::{BTreeMap, BTreeSet};
```
Comment (davidrusu) [inline src/merkle_reg.rs:5]: Let's keep the std/core imports together up top

File: src/merkle_reg.rs [hunk 2]
```diff
@@ -11,7 +11,7 @@ use crate::traits::{CmRDT, CvRDT};
 pub type Hash = [u8; 32];
 
 /// A node in the Merkle DAG
-#[derive(Debug, Clone, PartialEq, Eq)]
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, PartialOrd, Deserialize)]
```
Comment (davidrusu) [inline src/merkle_reg.rs:14]:
```suggestion
#[derive(Debug, Clone, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
```

File: src/merkle_reg.rs [hunk 3]
```diff
@@ -74,7 +74,7 @@ impl<'a, T> Content<'a, T> {
 /// The MerkleReg is a Register CRDT that uses the Merkle DAG
 /// structure to track the current value(s) held by this register.
 /// The leaves of the Merkle DAG are the current values.
-#[derive(Debug, PartialEq, Eq, Clone)]
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize, PartialOrd)]
```
Comment (davidrusu) [inline src/merkle_reg.rs:77]:
```suggestion
#[derive(Debug, Clone, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
```

File: src/merkle_reg.rs [hunk 4]
```diff
@@ -122,6 +122,21 @@ impl<T> MerkleReg<T> {
...
+                .iter()
+                .copied()
+                .filter_map(|leaf| self.dag.get(&leaf).map(|node| (leaf, node)))
+                .collect()
+        });
+
+        Content {
+            nodes: nodes.unwrap_or(BTreeMap::new()),
+        }
+    }
```
Comment (davidrusu) [inline src/merkle_reg.rs:138]: What do you think of moving this to into Content? We'd introduce a new ContentNode type for dealing with the contents of any single node, e.g. take a look at this sketch:

```rust
struct ContentNode<'a, T> {
    reg: &'a MerkleReg<T>,
    node: &'a Node<T>,
}

impl<'a> ContentNode<'a, T> {
    fn parents(&self) -> Content<T> {
        Content {
            reg: self.reg,
            nodes: self.node.parents.filter_map(|p| self.reg.node(p)).collect()
        }
    }

    fn value(&self) -> &T {
        &self.node.value
    }
}

struct Content<'a, T> {
   reg: &'a MerkleReg<T>
   nodes: BTreeMap<Hash, ContentNode<'a, T>>
}

impl Content {
    fn nodes() -> Vec<ContentNode<'a, T>> {
        self.nodes.values().collect()
    }
}
```
Then for this dag, the flow would be:
```
a  b
|\ |
| \|
c  d
 \/
  e
```
```rust
let content = reg.read().
content.values() // --> {a, b}
content.nodes() // -> [ContentNode(a_node), ContentNode(b_node) ]
content.nodes()[0].value() // -> a
content.nodes()[1].value() // -> b
content.nodes()[0].parents() // -> Content<T> at the time `a` was written
content.nodes()[0].parents().values() // -> { c, d }
content.nodes()[1].parents().values() // -> { d }
content.nodes()[0].parents().nodes()[0].value() // -> c
content.nodes()[0].parents().nodes()[0].parents() // -> Content at the time`c` was written
content.nodes()[0].parents().nodes()[0].parents().values() // -> {e}
content.nodes()[0].parents().nodes()[0].parents().nodes()[0].value() // -> e
content.nodes()[0].parents().nodes()[0].parents().nodes()[0].parents().values() // -> {}
```

Comment (bochaco) [reply src/merkle_reg.rs:138]: ok, looks neat, I'll add that now and push a new commit

**Review by davidrusu — APPROVED** [review]
:boat: Nice

**Comment by bochaco** [comment]
Superseded by PR #116

**Pull request closed without merging**